src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt

```diff
@@ -0,0 +1,235 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.refactoring
+
+import com.google.common.annotations.VisibleForTesting
+import com.intellij.openapi.actionSystem.DataContext
+import com.intellij.openapi.command.CommandProcessor
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Key
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiWhiteSpace
+import com.intellij.refactoring.BaseRefactoringProcessor
+import com.intellij.refactoring.RefactoringActionHandler
+import com.intellij.refactoring.RefactoringBundle
+import com.intellij.refactoring.ui.RefactoringDialog
+import com.intellij.refactoring.util.CommonRefactoringUtil
+import com.intellij.ui.components.JBTextField
+import com.intellij.ui.layout.panel
+import com.intellij.usageView.BaseUsageViewDescriptor
+import com.intellij.usageView.UsageInfo
+import com.intellij.usageView.UsageViewDescriptor
+import com.intellij.util.ui.JBUI
+import org.jetbrains.annotations.TestOnly
+import org.rust.ide.utils.GenericConstraints
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.openapiext.addTextChangeListener
+import org.rust.openapiext.isUnitTestMode
+import javax.swing.JComponent
+import javax.swing.JTextField
+
+class RsExtractTraitHandler : RefactoringActionHandler {
+    override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext?) {
+        val offset = editor.caretModel.offset
+        val element = file.findElementAt(offset) ?: return
+        val impl = element.ancestorOrSelf<RsImplItem>() ?: return
+        if (impl.traitRef != null) return
+        if (!CommonRefactoringUtil.checkReadOnlyStatus(project, impl)) return
+
+        val members = (impl.members ?: return).childrenOfType<RsItemElement>()
```

Comment:
I'm wondering if [inner attributes](https://doc.rust-lang.org/reference/items/implementations.html?highlight=impl#attributes-on-implementations) are covered here, could you confirm it or add a comment (and maybe a test)?

Reply:
Thanks, attributes were not supported, I fixed and added the test